### PR TITLE
gh-116935: Document that heap types need to support garbage collection

### LINKED ISSUE #116935: docs(easy): Document that heap types need to support garbage collection
# Documentation

In https://docs.python.org/3/c-api/typeobj.html#c.Py_TPFLAGS_HEAPTYPE, it does not mention that the heap type must support GC. However this is actually the case. The type itself needs to be visited by the GC because it forms a reference cycle with its own module object. We should document this.

### Linked PRs
* gh-118021
* gh-118092

### PR DESCRIPTION
**Pull request by savannahostrowski**
* Issue: gh-116935

----
📚 Documentation preview 📚: https://cpython-previews--118021.org.readthedocs.build/

Fixes: #116935

### CODE REVIEW

**Review by kalebers — APPROVED** [review]

**Comment by encukou** [comment]
Thank you for the addition!
Could you please add a link to [Supporting Cyclic Garbage Collection](https://docs.python.org/3/c-api/gcsupport.html#supporting-cyclic-garbage-collection), so readers can learn about the details if they don't already know them?

File: Doc/c-api/typeobj.rst [hunk 1]
```diff
@@ -1034,7 +1034,9 @@ and :c:data:`PyType_Type` effectively act as defaults.)
       the type, and the type object is INCREF'ed when a new instance is created, and
       DECREF'ed when an instance is destroyed (this does not apply to instances of
       subtypes; only the type referenced by the instance's ob_type gets INCREF'ed or
-      DECREF'ed).
+      DECREF'ed). Heap types must also support garbage collection as
+      they can form a `reference cycle <https://docs.python.org/3/c-api/gcsupport.html#supporting-cyclic-garbage-collection>`_
```
Comment (savannahostrowski) [inline Doc/c-api/typeobj.rst]: I was trying to way to add this link using elements/references but I couldn't figure it out. Maybe there's a nicer way to do this, open to suggestions!

Comment (JelleZijlstra) [reply Doc/c-api/typeobj.rst]: I think the wording is good but I would move the link to the words "support garbage collection". Also, you should be able to use a :ref: link, like:
```
:ref:`support garbage collection <supporting-cycle-detection>`
```
(Using the link target that points to the top of `gcsupport.rst`.)

Comment (savannahostrowski) [reply Doc/c-api/typeobj.rst]: Wow, I swear I tried that text for the reference link 😅 . I was also torn on where to place the link but I agree, it makes more sense where you suggested.

**Review by JelleZijlstra — APPROVED** [review]
The wording looks good but I'd like another review from someone who is more familiar with the C API.

**Review by kalebers — APPROVED** [review]
Well done, thanks for adding helpful information!

File: Doc/c-api/typeobj.rst [hunk 2]
```diff
@@ -1034,7 +1034,8 @@ and :c:data:`PyType_Type` effectively act as defaults.)
       the type, and the type object is INCREF'ed when a new instance is created, and
       DECREF'ed when an instance is destroyed (this does not apply to instances of
       subtypes; only the type referenced by the instance's ob_type gets INCREF'ed or
-      DECREF'ed).
+      DECREF'ed). Heap types must also :ref:`support garbage collection <supporting-cycle-detection>`
```
Comment (encukou) [inline Doc/c-api/typeobj.rst]: Let's soften the language. It's technically possible to create a type that doesn't form reference cycles, if you're really determined. (AFAIK: it'd have to be an immutable non-subclassable type with no associated module and with instances whose attributes can't form cycles.)
```suggestion
      DECREF'ed). Heap types should also :ref:`support garbage collection <supporting-cycle-detection>`
```

**Review by encukou — APPROVED** [review]

**Comment by savannahostrowski** [comment]
Thanks for the reviews, folks!

**Pull request merged**